### DISCUSSION TITLE
Improvements for testing

### DISCUSSION THREAD

**Discussion by VampireAotD** [westacks/telebot#87] [Ideas]
Hello!

Recently, i wanted to test some of my handlers, and discovered that some of them are not giving me expected results.
For example, wanted to test a handler that returns user a message with caption and photo:
```php
public function testExampleHandler(): void
{
   // Just for example
   $this->bot->fake()->sendPhoto([
       'chat_id' => -1,
       'caption' => $this->faker->sentence,
       'photo'   => $this->faker->imageUrl,
   ]);
}
```

But got this exception: `Cannot cast value of type string to type WeStacks\TeleBot\Objects\PhotoSize at vendor/westacks/telebot/src/Contracts/TelegramObject.php:40`.

This was a bit confusing, because if not faking the request, everything works perfectly. So i went to check what is wrong, and i discovered that i can only pass `chat_id`, everything else is pretty much hardcoded in `mock` method:

```php
// SendPhotoMethod mock method
public function mock($arguments)
    {
        return new Message([
            'message_id' => mt_rand(1, 100),
            'from' => [
                'id' => mt_rand(1, 100),
                'is_bot' => false,
                'first_name' => '',
                'last_name' => '',
                'username' => '',
                'language_code' => '',
            ],
            'chat' => [
                'id' => $arguments['chat_id'],
                'type' => 'private',
                'title' => '',
                'username' => '',
                'first_name' => '',
                'last_name' => '',
                'all_members_are_administrators' => false,
            ],
            'date' => time(),
            'photo' => [
                'file_id' => '',
                'width' => mt_rand(1, 100),
                'height' => mt_rand(1, 100),
                'file_size' => mt_rand(1, 100),
            ],
        ]);
    }
```

And in other handlers as well, for example in `ExportChatInviteLinkMethod`:
```php
public function mock($arguments)
{
    return 'https://telegram.me/joinchat/test';
}
```

But  mostly in all tests i need to check some specific data, like `photo`, or `first_name`, which are hardcoded, but every method has `$arguments` which are mostly not used.

So the question is: **do you plan to add support of using passed arguments in tests?**

Maybe, for example, adding `defaults` method to `TelegramMethod` class will help? That way each method will be able to return default values for mocks or override defaults if user provided arguments himself, something like this:
```php
abstract class TelegramMethod
{
    // ...
    
    /**
     * Mocked data for testing.
     */
    abstract protected function defaults(): mixed;
    
    // ...
}
```

```php
class SendPhotoMethod extends TelegramMethod
{
    //...
    
    protected function defaults(): mixed
    {
        return [
            'message_id' => mt_rand(1, 100),
            'from' => [
                'id' => mt_rand(1, 100),
                'is_bot' => false,
                'first_name' => '',
                'last_name' => '',
                'username' => '',
                'language_code' => '',
            ],
            'chat' => [
                'id' => $arguments['chat_id'],
                'type' => 'private',
                'title' => '',
                'username' => '',
                'first_name' => '',
                'last_name' => '',
                'all_members_are_administrators' => false,
            ],
            'date' => time(),
            'photo' => [
                'file_id' => '',
                'width' => mt_rand(1, 100),
                'height' => mt_rand(1, 100),
                'file_size' => mt_rand(1, 100),
            ],
        ];
    }

    public function mock($arguments)
    {
        $data = $this->defaults();

        if (is_array($arguments)) {
            $data = array_merge($data, $arguments);
        }

        return new Message($data);
     }
}
```

```php
class ExportChatInviteLinkMethod extends TelegramMethod
{
    //...
    
    protected function defaults(): mixed
    {
        return 'https://telegram.me/joinchat/test';
    }

    public function mock($arguments)
    {
        return $arguments ?? $this->defaults();
    }
}
```

I'm sorry if i wrote this in wrong category, just wanted to know if this could be possible?

Also, thank you for your work on this project, its always nice to see new updates :)

**Comment by punyflash** [comment]
Hi! In general testing Telegram API a bit complicated as there is a bunch of logic happens inside Telegram servers itself like invoice generation, callbacks, inline callbacks etc. And all attached to the IDs generated by Telegram. I was thinking for two years how to make tests more convenient, so they could replicate logic of Telegram API easy and be a verification that if your code runs it should run with Telegram servers as well. But I didn't find a good solution.

I'm not satisfied with the current way of mocking responses, yeah, but I've not came up with anything better, so the topic of testing is open for discussion.

**Comment by VampireAotD** [comment]
Hello!

Sorry for the late response.

The current implementation of mocks is indeed effective and in many situations mocks Telegram's responses well, I think, I wouldn't came up with a better solution myself. However, the main aspect of this discussion is not to completely rethink current implementation of mocks, but to slightly extend them and adapt them to be more flexible for testing.

As I wrote earlier, many methods have already hardcoded data, and in some cases this limits the completeness of testing, as it is impossible to check whether my data will actually be sent to Telegram.

In my previous post, I tried to provide a solution that I think could be used to improve testing. In your opinion, would it be possible to add a similar solution to allow mocks to work with data that user provided?
